requestmanager/executor/executor.go

```diff
@@ -18,183 +19,172 @@ import (
+		// check for immediate result
 		select {
 		case result = <-resultChan:
 		default:
-			err := re.sendRestartAsNeeded()
-			if err != nil {
+			// if no immediate result
+			// initiate remote request if not already sent (we want to fill out the doNotSendCids on a resume)
+			if err := onlyOnce.startRemoteRequest(); err != nil {
 				return err
 			}
+			// wait for block result
 			select {
-			case <-re.ctx.Done():
+			case <-rt.Ctx.Done():
 				return ipldutil.ContextCancelError{}
 			case result = <-resultChan:
 			}
 		}
-		err = re.processResult(traverser, lnk, result)
-		if _, ok := err.(hooks.ErrPaused); ok {
-			err = re.waitForResume()
-			if err != nil {
-				return err
-			}
-			err = traverser.Advance(bytes.NewBuffer(result.Data))
-			if err != nil {
-				return err
-			}
-		} else if err != nil {
+		log.Debugf("successfully loaded link=%s, nBlocksRead=%d", lnk, rt.Traverser.NBlocksTraversed())
+		// advance the traversal based on results
+		err = e.advanceTraversal(rt, result)
+		if err != nil {
 			return err
 		}
-	}
-}
 
-func (re *requestExecutor) run() {
-	err := re.traverse()
-	if err != nil {
-		if !isContextErr(err) {
-			select {
-			case <-re.ctx.Done():
-			case re.inProgressErr <- err:
-			}
-		}
-	}
-	select {
-	case terminalError := <-re.terminalError:
-		select {
-		case re.inProgressErr <- terminalError:
-		case <-re.env.Ctx.Done():
+		// check for interrupts and run block hooks
+		err = e.processResult(rt, lnk, result)
+		if err != nil {
+			return err
 		}
-	default:
 	}
-	re.terminateRequest()
-	close(re.inProgressChan)
-	close(re.inProgressErr)
 }
 
-func (re *requestExecutor) sendRequest(request gsmsg.GraphSyncRequest) {
-	re.env.SendRequest(re.p, request)
-}
-
-func (re *requestExecutor) terminateRequest() {
-	re.env.TerminateRequest(re.request.ID())
+func (e *Executor) processBlockHooks(p peer.ID, response graphsync.ResponseData, block graphsync.BlockData) error {
+	result := e.blockHooks.ProcessBlockHooks(p, response, block)
+	if len(result.Extensions) > 0 {
+		updateRequest := gsmsg.UpdateRequest(response.RequestID(), result.Extensions...)
+		e.manager.SendRequest(p, updateRequest)
+	}
+	return result.Err
 }
 
-func (re *requestExecutor) runBlockHooks(blk graphsync.BlockData) error {
-	response := re.lastResponse.Load().(gsmsg.GraphSyncResponse)
-	return re.env.RunBlockHooks(re.p, response, blk)
+func (e *Executor) onNewBlock(rt RequestTask, block graphsync.BlockData) error {
+	rt.DoNotSendCids.Add(block.Link().(cidlink.Link).Cid)
+	response := rt.LastResponse.Load().(gsmsg.GraphSyncResponse)
+	return e.processBlockHooks(rt.P, response, block)
 }
 
-func (re *requestExecutor) waitForResume() error {
-	select {
-	case <-re.ctx.Done():
-		return ipldutil.ContextCancelError{}
-	case re.pendingExtensions = <-re.resumeMessages:
-		re.restartNeeded = true
-		return nil
+func (e *Executor) advanceTraversal(rt RequestTask, result types.AsyncLoadResult) error {
+	if result.Err != nil {
```

Comment:
could result.Err also be a `traversal.SkipMe`?

Reply:
Probably not.